Mbed OS 5.15.8, updated sensor drivers & config

JanneKiiskila opened the pull request:
* Add network IP address & MAC printing.
* Update Mbed OS to known release version from 5.15-branch.
* Update sensor driver code to latest version.
* Change sensor abstraction level to be it's own. 
    * It does not depend on the board target - you can attach the sensor board to almost any STM32 devkit and it will work.
    *Tidy up the way the configurations are done - centralize them to the header to minimize ifdefs.

With the default option of `SENSOR_IKS01A2=0` they get left out completely. Compiler knows how to optimize them out.

```
| Module                      |           .text |       .data |          .bss |
|-----------------------------|-----------------|-------------|---------------|
| [fill]                      |       648(+648) |     14(+14) |       92(+92) |
| [lib]/c.a                   |   49008(+49008) | 2474(+2474) |       89(+89) |
| [lib]/gcc.a                 |     7896(+7896) |       0(+0) |         0(+0) |
| [lib]/m.a                   |     1032(+1032) |       1(+1) |         0(+0) |
| [lib]/misc                  |       180(+180) |       4(+4) |       28(+28) |
| [lib]/nosys.a               |         32(+32) |       0(+0) |         0(+0) |
| [lib]/stdc++.a              |     5928(+5928) |       8(+8) |       44(+44) |
| anjay-mbedos/Anjay          | 124058(+124058) |       0(+0) |         0(+0) |
| anjay-mbedos/avs_commons    |   26844(+26844) |     12(+12) |       16(+16) |
| anjay-mbedos/src            |   12606(+12606) |       4(+4) |     158(+158) |
| conn_monitoring_object.o    |       904(+904) |       0(+0) |       92(+92) |
| device_config_serial_menu.o |     5604(+5604) |       0(+0) |         0(+0) |
| device_object.o             |     1192(+1192) |       0(+0) |       92(+92) |
| main.o                      |     2488(+2488) |       0(+0) |     828(+828) |
| mbed-os/components          |       282(+282) |       0(+0) |         0(+0) |
| mbed-os/drivers             |       720(+720) |       0(+0) |         0(+0) |
| mbed-os/events              |     1454(+1454) |       0(+0) |   3108(+3108) |
| mbed-os/features            | 166262(+166262) |   234(+234) | 23641(+23641) |
| mbed-os/hal                 |     2682(+2682) |       8(+8) |     247(+247) |
| mbed-os/platform            |     5704(+5704) |   276(+276) |     462(+462) |
| mbed-os/rtos                |   11252(+11252) |   168(+168) |   5981(+5981) |
| mbed-os/targets             |   12914(+12914) |       5(+5) |     730(+730) |
| serial_menu.o               |       494(+494) |       0(+0) |         0(+0) |
| Subtotals                   | 440184(+440184) | 3208(+3208) | 35608(+35608) |
Total Static RAM memory (data + bss): 38816(+38816) bytes
Total Flash memory (text + data): 443392(+443392) bytes

Update Image: ./BUILD/NUCLEO_F429ZI/GCC_ARM/Anjay-mbedos-client_update.bin
Image: ./BUILD/NUCLEO_F429ZI/GCC_ARM/Anjay-mbedos-client.bin

```

Commit:
Use mbed-os-5.15.8

Stick to an official release of the Mbed OS 5.15-branch (.8 is
the latest one). The current Anjay will not compile against Mbed OS 6.x.

Commit:
Use latest for X_NUCLEO_IKS01A2 library

It has been recently updated, giving for example Mbed OS 6
compatibility (if one would venture to that direction).

Commit:
Add IP address & MAC printing

This is quite useful information when troubleshooting network issues
sometimes.

Update NOTICE years, it is year 2022 already.

Commit:
Add error print if default nw i/f not working

We should probably tell if it's failing, so that it's easier to debug.
There can be boards w/o any default network interface.

Review comment covered - \n added to error print.

Commit:
README.md - editorial changes

Tidy up the README.md just a bit.

Commit:
README.md - update Mbed OS link

The URL gets redirected, change it to the direct end target.

Commit:
Re-structure IKS01A2 sensor configuration

The sensor board has nothing to do with the TARGET_DISCO_L496AG per se,
you can add that same sensor board to literally almost any STM32 target.
Thus, it needs to have it's own configuration.

Tidying up also the code - we do not need to have #ifdefs everywhere.
We just specify the config in the mbed_app.json and have "dummy"
empty implementations in the header files that do nothign - the compiler
actually optimizes them out completely (even if they are "called" in
main.cpp). This is how it is usually done in the Linux world and
this allows for "neater" code flow as it does not require to any

I.e.

```C
        if (setup_security_object() || setup_server_object()
            || device_object_install(anjay)
            || joystick_object_install(anjay)
            || humidity_object_install(anjay)
            || barometer_object_install(anjay)
            || magnetometer_object_install(anjay)
            || accelerometer_object_install(anjay)

```
This split for the configs wit #ifdefs would mean we would
have code like below:

```C
        if (setup_security_object() || setup_server_object()
            || device_object_install(anjay)
            || joystick_object_install(anjay)
            || humidity_object_install(anjay)
            || barometer_object_install(anjay)
            || magnetometer_object_install(anjay)
            || accelerometer_object_install(anjay)
```

We have multiple places where that happens. We centralize it
all to the header file instead.

kFYatek commented on main.cpp:
```diff
@@ -434,6 +432,15 @@ class NetworkService {
             NETWORK->set_access_technology(config.modem_config.rat);
         }
 
+        // Print IP address and MAC address, quite useful in troubleshooting
+        err = netif->get_ip_address(&sa);
+        if (err != NSAPI_ERROR_OK) {
+            printf("get_ip_address() - failed, status %d\n", err);
+        } else {
+            printf("IP: %s\n", (sa.get_ip_address() ? sa.get_ip_address() : "None"));
+            printf("MAC address: %s\n", (netif->get_mac_address() ? netif->get_mac_address() : "None"));
+        }
```
Nitpick: `\r\n` is a more idiomatic line ending when writing to serial port. It actually makes a difference in some terminal software (e.g. `picocom`).

JanneKiiskila replied:
It should not - https://github.com/AVSystem/Anjay-mbedos-client/blob/master/mbed_app.json#L17 - that should automatically handle the `\n\r` issue.

kFYatek replied:
Oh, you're right. I got it mixed up due to working on other projects without such translation concurrently.

kFYatek commented on main.cpp:
```diff
@@ -411,8 +405,12 @@ class NetworkService {
+
         NetworkInterface *netif = NetworkInterface::get_default_instance();
         if (!netif) {
+            printf("ERROR - can't get default network instance!");
```
Adding a line ending (`\r\n`) would be nice.

JanneKiiskila replied:
Duly noted, original commit fixed and force pushed.

kFYatek commented on accelerometer.h:
```diff
@@ -19,14 +19,25 @@
-#endif // TARGET_DISCO_L496AG
+#else // Define dummy functions, compiler optimizes out
+
+int accelerometer_object_install(anjay_t *anjay) {
+    return 0;
+}
+void accelerometer_object_uninstall(anjay_t *anjay) {
+}
+
+void accelerometer_object_update(anjay_t *anjay) {
+}
```
These (and similar declarations in other headers) should either be moved to the `.cpp` files, or declared as `static inline`, or put into an unnamed namespace. Otherwise, link errors may occur if this header is included from multiple translation units.

JanneKiiskila replied:
Static inline seems to be how to do it Linux. I'll do that.

JanneKiiskila replied:
Done.

Commit:
Make all dummy functions static inline

Make them static inline based on review comments in
https://github.com/AVSystem/Anjay-mbedos-client/pull/11
from kFYatek (GitHub ID).

kFYatek commented:
Looks good, thanks!
